drivers/dma/dma_mcux_edma.c

```diff
@@ -507,29 +614,119 @@ static int dma_mcux_edma_reload(const struct device *dev, uint32_t channel,
 		goto cleanup;
 	}
 
-	/* If the tcdPool is not in use (no s/g) then only a single TCD can be active at once. */
-	if (data->busy && data->edma_handle.tcdPool == NULL) {
-		LOG_ERR("EDMA busy. Wait until the transfer completes before reloading.");
-		ret = -EBUSY;
-		goto cleanup;
-	}
+	if (data->transfer_settings.cyclic) {
+		if (data->transfer_settings.empty_tcds == 0) {
+			LOG_DBG("TCD list is full in loop mode.");
+			ret = -ENOBUFS;
+			goto cleanup;
+		}
 
-	EDMA_PrepareTransfer(
-		&(data->transferConfig),
-		(void *)src,
-		data->transfer_settings.source_data_size,
-		(void *)dst,
-		data->transfer_settings.dest_data_size,
-		data->transfer_settings.source_burst_length,
-		size,
-		data->transfer_settings.transfer_type);
-
-	const status_t submit_status =
+		/* Convert size into major loop count */
+		size = size/data->transfer_settings.dest_data_size;
+
+		/* Append the transfer normally */
+		tcd = &(DEV_CFG(dev)->tcdpool[channel][data->transfer_settings.write_idx]);
+		pre_tcd = &(DEV_CFG(dev)->tcdpool[channel][(data->transfer_settings.write_idx - 1)%CONFIG_DMA_TCD_QUEUE_SIZE]);
+#if (CONFIG_DMA_MCUX_EDMA || CONFIG_DMA_MCUX_EDMA_V3)
+		tcd->SADDR = src;
+		tcd->DADDR = dst;
+		tcd->BITER = size;
+		tcd->CITER = size;
+		/* Enable automatically stop */
+		tcd->CSR |= DMA_CSR_DREQ(1U);
+		sw_id = tcd->DLAST_SGA;
+#elif CONFIG_DMA_MCUX_EDMA_V4
+		EDMA_TCD_SADDR(tcd,kEDMA_EDMA4Flag) = src;
+		EDMA_TCD_DADDR(tcd,kEDMA_EDMA4Flag) = dst;
+		EDMA_TCD_BITER(tcd,kEDMA_EDMA4Flag) = size;
+		EDMA_TCD_CITER(tcd,kEDMA_EDMA4Flag) = size;
+		/* Enable automatically stop */
+		EDMA_TCD_CSR(tcd,kEDMA_EDMA4Flag) |= DMA_CSR_DREQ(1U);
+		sw_id = EDMA_TCD_DLAST_SGA(tcd,kEDMA_EDMA4Flag);
+#endif
+
+		/* Manaually stop DMA whatever it is running or already stopped(Automatically)
+		   Make sure the code between EDMA_DisableChannelRequest and EDMA_EnableChannelRequest is minimum.
+		*/
+		EDMA_DisableChannelRequest(DEV_BASE(dev),channel);
+#ifdef CONFIG_DMA_MCUX_EDMA
+		while(DEV_BASE(dev)->TCD[channel].CSR & DMA_CSR_ACTIVE_MASK)
+#elif (CONFIG_DMA_MCUX_EDMA_V3 || CONFIG_DMA_MCUX_EDMA_V4)
+		while(DEV_BASE(dev)->CH[channel].TCD_CSR & DMA_CH_CSR_ACTIVE_MASK)
```

Comment:
I'm not sure I'm following the logic here- why do we wait until the EDMA is not active here? The `dma_reload` function should simply queue a new buffer into the EDMA (which occurs above) then exit, right? There is no requirement to wait for the DMA to complete execution here.

It seems like at this point we should:
- clear the DREQ bit on the prior TCD, so the next TCD can be loaded
- check the next TCD address, and restart the DMA if `hw_id == sw_id` (as this indicates the TCD we just configured was loaded already, right?)
- otherwise, set the ESG bit to resolve the HW bug, and make sure the DREQ bit is set on the TCD we are queueing

I'm not clear why we need to disable channel requests or wait for the EDMA to be inactive. Shouldn't it be sufficient to configure the TCD, enable the EDMA to load the TCD, then perform the safety check to make sure the EDMA did not load an invalid TCD?

Reply:
If we do not stop DMA during this "appending procedure", HW ID may change at any time, that is to say, we don't know exactly which TCD DMA is working on.  DMA may have finished all transfers, may be transferring the last one, may be transferring the others, all handling of these conditions are different.
By disable the channel request, we cut off the signal from the peripheral, no more transfer request. DMA stops after current burst finished. Why we wait for DMA to be inactive is because even we cut off the transfer request, there may be a burst transfer is on-going, DMA has its own internal buffer and logics. DMA may modify the same HW reg while CPU is trying to make a modification.

Reply:
Ok, that makes sense. Isn't there a risk here though of "drops" in transfers? For example with audio, wouldn't the DMA being inactive here manifest as a momentary pause in the I2S audio stream and therefore a pause in the audio from the codec?

Reply:
Risk of "drops":
that's why we need make this part of code as short as possible. The longest path of code between EDMA_DisableChannelRequest and EDMA_EnableChannelRequest is in case DMA has finished all transfers and we need rewrite HW regs. Even in this case, all codes to execute are about 10 simple instructions(set value, comparing).  It should be very quick.  On RT1170, SAI has 16 bytes FIFO, given the sample rate: 44K, 2 bytes word, 2 channels, it takes about 90us to full the FIFO.  I think it is much far longer than the execution time of the critical code.

last TCD transfering:
In this case, hw_id should be current writing TCD address, we have code to handle this case:
				/* DMA is running on last transfer. HW has loaded the last one,
				 * we need ensure it's DREQ is cleared.
				 */
				EDMA_EnableAutoStopRequest(DEV_BASE(dev), channel, false);
				LOG_DBG("Last transfer.");_